esphome/components/samsung/samsung.h

```diff
@@ -0,0 +1,228 @@
+#pragma once
+
+#include "esphome/components/climate/climate_mode.h"
+#include "esphome/components/climate_ir/climate_ir.h"
+
+namespace esphome {
+namespace samsung {
+
+#define GETBITS8(data, offset, size) (((data) & (((uint8_t) UINT8_MAX >> (8 - (size))) << (offset))) >> (offset))
+
+static const uint32_t SAMSUNG_IR_FREQUENCY_HZ = 38000;
+static const uint32_t SAMSUNG_AIRCON1_HDR_MARK = 3000;
+static const uint32_t SAMSUNG_AIRCON1_HDR_SPACE = 9000;
+static const uint32_t SAMSUNG_AIRCON1_BIT_MARK = 500;
+static const uint32_t SAMSUNG_AIRCON1_ONE_SPACE = 1500;
+static const uint32_t SAMSUNG_AIRCON1_ZERO_SPACE = 500;
+static const uint32_t SAMSUNG_AIRCON1_MSG_SPACE = 2000;
+
+const uint16_t K_SAMSUNG_AC_EXTENDED_STATE_LENGTH = 21;
+const uint16_t K_SAMSUNG_AC_SECTION_LENGTH = 7;
+
+// Temperature
+const uint8_t K_SAMSUNG_AC_MIN_TEMP = 16;  // C   Mask 0b11110000
+const uint8_t K_SAMSUNG_AC_MAX_TEMP = 30;  // C   Mask 0b11110000
+
+// Mode
+const uint8_t K_SAMSUNG_AC_AUTO = 0;
+const uint8_t K_SAMSUNG_AC_COOL = 1;
+const uint8_t K_SAMSUNG_AC_DRY = 2;
+const uint8_t K_SAMSUNG_AC_FAN = 3;
+const uint8_t K_SAMSUNG_AC_HEAT = 4;
+
+// Fan
+const uint8_t K_SAMSUNG_AC_FAN_AUTO = 0;
+const uint8_t K_SAMSUNG_AC_FAN_LOW = 2;
+const uint8_t K_SAMSUNG_AC_FAN_MED = 4;
+const uint8_t K_SAMSUNG_AC_FAN_HIGH = 5;
+
+// Swing
+const uint8_t K_SAMSUNG_AC_SWING_V = 0b010;
+const uint8_t K_SAMSUNG_AC_SWING_H = 0b011;
+const uint8_t K_SAMSUNG_AC_SWING_BOTH = 0b100;
+const uint8_t K_SAMSUNG_AC_SWING_OFF = 0b111;
+
+// Power
+const uint8_t K_NIBBLE_SIZE = 4;
+const uint8_t K_LOW_NIBBLE = 0;
+const uint8_t K_HIGH_NIBBLE = 4;
+
+static const uint8_t K_RESET[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH] = {0x02, 0x92, 0x0F, 0x00, 0x00, 0x00, 0xF0,
+                                                                    0x01, 0x02, 0xAE, 0x71, 0x00, 0x15, 0xF0};
+
+/// Native representation of a Samsung A/C message.
+union SamsungProtocol {
+  uint8_t raw[K_SAMSUNG_AC_EXTENDED_STATE_LENGTH];  ///< State in code form.
+  struct {                                          // Standard message map
+    // Byte 0
+    uint8_t : 8;
+    // Byte 1
+    uint8_t : 4;
+    uint8_t : 4;  // Sum1Lower
+    // Byte 2
+    uint8_t : 4;  // Sum1Upper
+    uint8_t : 4;
+    // Byte 3
+    uint8_t : 8;
+    // Byte 4
+    uint8_t : 8;
+    // Byte 5
+    uint8_t : 4;
+    uint8_t Sleep5 : 1;
```

Comment:
> Function, method and variable names are `lower_snake_case`

https://esphome.io/guides/contributing#codebase-standards
```suggestion
    uint8_t sleep_5 : 1;
```